cache/contenthash/checksum.go

```diff
@@ -928,31 +928,64 @@ func (cc *cacheContext) checksum(ctx context.Context, root *iradix.Node[*CacheRe
 	return cr2, true, nil
 }
 
+// pathSet is a set of path prefixes that can be used to see if a given path is
+// lexically a child of any path in the set. All paths provided to this set
+// MUST be absolute and use / as the separator.
+type pathSet struct {
+	// prefixes contains paths of the form "/a/b/", so that we correctly detect
+	// /a/b as being a parent of /a/b/c but not /a/bc.
+	prefixes []string
+}
+
+// add a path to the set. This is a no-op if includes(path) == true.
+func (s *pathSet) add(p string) {
+	// Ensure the path is absolute and clean.
+	p = path.Join("/", p)
+	if !s.includes(p) {
+		if p != "/" {
+			p += "/"
+		}
+		s.prefixes = append(s.prefixes, p)
+	}
+}
+
+// includes returns true iff there is a path in the pathSet which is a lexical
+// parent of the given path. The provided path MUST be an absolute path and
+// MUST NOT contain any ".." components, as they will be path.Clean'd.
+func (s pathSet) includes(p string) bool {
+	// Ensure the path is absolute and clean.
+	p = path.Join("/", p)
+	if p != "/" {
+		p += "/"
+	}
+	for _, prefix := range s.prefixes {
+		if strings.HasPrefix(p, prefix) {
+			return true
+		}
+	}
+	return false
+}
+
 // needsScan returns false if path is in the tree or a parent path is in tree
 // and subpath is missing.
 func (cc *cacheContext) needsScan(root *iradix.Node[*CacheRecord], path string, followTrailing bool) (bool, error) {
 	var (
-		lastGoodPath    string
+		goodPaths       pathSet
 		hasParentInTree bool
 	)
 	k := convertPathToKey(path)
 	_, cr, err := getFollowLinksCallback(root, k, followTrailing, func(subpath string, cr *CacheRecord) error {
+		// If we found a path that exists in the cache, add it to the set of
+		// known-scanned paths. Otherwise, verify whether the not-found subpath
+		// is inside a known-scanned path (we might have hit a "..", taking us
+		// out of the scanned paths, or we might hit a non-existent path inside
+		// a scanned path). getFollowLinksCallback iterates left-to-right, so
```

Comment:
We could improve the comment for `getFollowLinksCallback` a bit to make it clear in what conditions the callback will get called. Afaics the semantics changed in this PR a bit, but there is no definition to clarify that.

Reply:
Hmm, I though about what I could add but the semantics are unchanged -- it's just that we do more cache lookups than we used to (we were skipping doing cache lookups we weren't supposed to). The callback is still only called after a cache lookup (except for the first verbatim lookup).

Reply:
Smth like, callback first gets called for the root path, then for every individual component in that path, if that component is a symlink then that is resolved and every new component is passed to callback again.

Reply:
I came up with:

```
// cb is a callback that is called for each path component encountered during
// path resolution (after the path component is looked up in the cache). This
// means for a path like /a/b/c, the callback will be called for at least
//
//	{/, /a, /a/b, /a/b/c}
//
// Note that if any of the components are symlinks, the paths will depend on
// the symlink contents and there will be more callbacks. If the requested key
// has a trailing slash, the callback will also be called for the final
// trailing-slash lookup (/a/b/c/ in the above example). Note that
// getFollowLinksCallback will try to look up the original key directly first
// and the callback is not called for this first lookup.
```

It's hard to explain the exact semantics without just describing how path lookups work (and just talking about cache lookups meant there was less need to explain things) but this should be a bit easier to understand if you don't know how lookups work.

Reply:
SGTM

Reply:
Pushed :smile_cat: